Update the proposal for stage 2 (#12)
tc39/proposal-intl-duration-format: README.md

## Motivation

* Users need all types of duration format depending on the requirements of their application. For example, to show how long a flight takes, the duration should be in Short or Narrow format
  - 1 hr 40 min 60 sec → Short 
  - 1 h 40 m 60 sec  → Narrow

> ## Requirements & Design

Reviewer: I'm not sure I've seen such a section in an explainer before, but I have to admit that it seems useful. Perhaps we should move it into it's own file called `requirements.md` or something?

Reviewer 2: I think it's fine to keep in the README, especially at Stage 2.

Author: Okay, I will leave it for now, and we can discuss updating it before Stage 3.

In this section, we are going to illustrate each user needs (requirements) and a design for each need (requirement)

### Input Value

>   * Users needs to identify the input values

Reviewer: I'm not sure I understand what "identifying" input values entails. Perhaps a longer description would help here?

Author: okay, I added more descriptions

  * For example
    * format( 100 seconds)
    * format( 1 hour, 20 minutes)

#### Design

>   * Input value will be an object from `Temporal.Duration`

Reviewer: ```suggestion
  * Input value will be an object of type `Temporal.Duration`
```

  * Example:

>      * `const d = new DurationFormat();`

Reviewer: Would we still need to pass a locale argument?

Reviewer 2: Intl defaults to the system locale if you don't pass one.

>      * `d.format(new Temporal.Duration({hour: 3, minute: 4});`

Reviewer: ```suggestion
     * `d.format(new Temporal.Duration.from({hours: 3, minutes: 4});`
```

### Formatting width
Users want to determine several types of the formatting width as following
  | Format width  | Example               |
  | -----------   | -----------           |
  |  Wide         |1 hour and 50 minutes  |
  |  Short        |1 hr, 50 min           |
  |  Narrow       |1h 50 m                |

>   |  Numeric      |                       |

Reviewer 2: Please fill this in.  This is important information for Stage 2.

Reviewer: For now, I've excluded this option from the spec since it seems unclear to me what this option would look like.

Author: I have removed it, it is only exit in the time format

  |  Dotted       |1: 50: 00              |

#### Design

>   * The user can determine the formatting width using a  parameter `style` and the value of this parameter take a ` string` as following:

Reviewer: ```suggestion
  * The user can determine the formatting width using a `style` parameter which accepts the following string values:
```

>       * “wide” 

Reviewer 2: Please make the names string literals: \`"wide"\` (rendered as `"wide"`)

Reviewer: I propose that we use `"long"` instead for consistency with ECMA 402 (`DateTimeFormat`, `RelativeTimeFormat`, `ListFormat`) since I'm not even sure if consistency with ICU is necessary.

Reviewer 2: +1 on long/short/narrow instead of wide/short/narrow

Author: Done !

      * “short”
      * “narrow”
      * “numeric” 
      * “dotted”

### Supported Time Fields
Users need the following fields to be supported
  * Year
  * Month
  * Week
  * Day
  * Hour
  * Minute
  * Second
  * Millisecond
  * Microsecond
  * Nano Second

#### Design

> We are going to support the following fields as a beginning

Reviewer 2: I don't understand this section.  You have a longer list up above, which is supported by Temporal.

Reviewer: Me neither. I propose we remove this section altogether and just mention somewhere that we will support all `Temporal.Duration` fields since we've received consensus on adding weeks in the champions group anyway.

Author: This is an issue that we are going to discuss it next week. We are going to discuss if we are going to support the same fields as `Temporal.Duration` or we are going to determine a set of fields to support as a start.

  * Week
  * Day
  * Hour
  * Minute
  * Second
  * Millisecond
  * Microsecond
  * Nano Second

### Determining Time Fields
Users need to determine the time fields,
 * For example: 6059 seconds
   * Hours, Minutes, Seconds
     * 1 hours, 40 minutes and  59 seconds
   * Minutes and Seconds
     * 100 minutes and 59 seconds.
 

> #### Design

Reviewer 2: Are you suggesting that we perform arithmetic here?  That's a very important detail.  If you want to perform arithmetic (allow the user to pass 6059 seconds directly into Intl.DurationFormat and have it render "100 minutes and 59 seconds"), you should explain the algorithm used to perform that arithmetic.

Author: That is correct, let us also to postpone writing the algorithm until the discussion next week.

   * Users can determine the time fields using `LargestField` and `SmallestField` parameters.
   * Example: 
   ```javascript
    {
      LargestField: "hour",
      SmallestField: "second"
    }
  ```
   * Default values
    ```javascript
         LargestField: "Day",
         SmallestField: "second"
    ```
### Hide zero-value fields
Users needs to hide zero-value fields depends in some criteria
  * Hide all zero-value fields
    * Hours, Minutes, Seconds (without the `hide` option)
      * 1 hours, 0 minutes and 59 seconds.
    * Hours, Minutes, Seconds (with the `hide` option)
      * 1 hours and 59 seconds.
  * Hide-upper-limit with zero-valued.
    * Hours, Minutes, Seconds, MilliSeconds (without the `hide-upper-limit` option)
      * 0 hr, 20 min, 0 sec, 100 msec .
    * Hours, Minutes, Seconds MilliSeconds(with the `hide-upper-limit` option)
      * 20 min, 0 sec, 100 msec.
  * Hide-lower-limit with zero-valued.

#### Design
- Users can set the `HidingZeroValuedOption` parameter  by one of the following values:

>     * “ALL”, // Hide all the fields that have a zero-valued.

Reviewer 2: Please use \`"all"\` syntax.

Author: Do you want also to use only small letters ?

    * “LEADING_AND_TRAILING”, // Hide all the zero fields in the leading or the training
    * “LEADING_ONLY”,
    * “TRAILING_ONLY”, 
    * “NONE”, // Do not hide any zero-valued fields.

- Default Value: “NONE”

### Locale aware format
 - Users needs the formatting to be dependent on the locale
 - For example:
    * en-US
      * 1 hour, 46 minutes and 40 seconds
    * fr-FR
      * 1 heure, 46 minutes et 40 secondes

## API design

> I propose to have a class called `DurationFormat` 

Reviewer: This line is unnecessary.

Author: Done !

  ```javascript
    const formatter =        
        new Intl.DurationFormat({

>                               locale: 'en-US',

Reviewer 2: The locale is always the first argument.  Also, your casing conventions are not correct.

I think you would want:

```javascript
const formatter = new Intl.DurationFormat('en-US', {
    largestField: 'hour',
    smallestField: 'second', 
		style: 'short',
		round: true,
		hidingZeroValuedOption: NONE,
});
```

As stated before, I'm not 100% onboard with this API, but it's good enough for Stage 2.

Reviewer: +1 on locale as the first argument. Regarding the options, I thought we had decided to use the more explicit `"fields"` option for Stage 2 and iterate on these solutions before Stage 3, so that's what I did in the spec.

Author: Done !

		                		      largestField: 'hour',
  					                  smallestField: 'second', 
					                    style: 'short',

> 					                    round: true,

Reviewer 2: What does `round: true` mean?  That's not documented anywhere.

Author: I have added it.

					                    HidingZeroValuedOption: NONE,
                              });

    const duration = 
        Temporal.Duration.from({
            hours: 2, minutes: 46, seconds: 40
         });

    console.log(formatter.format(duration));

>       // output: 2 hr 46 min 40 sec

Reviewer: I propose that we move this example to somewhere close to the top of this document for clarity.

### Formatting width
Users want to determine several types of the formatting width as following
  | Format width  | Example               |
  | -----------   | -----------           |
  |  Wide         |1 hour and 50 minutes  |
  |  Short        |1 hr, 50 min           |
  |  Narrow       |1h 50 m                |
  |  Dotted       |1: 50: 00              |

#### Design
  * The user can determine the formatting width using a  parameter `style` and the value of this parameter take a ` string` as following:
      * `“wide”`
      * `“short”`
      * `“narrow”`
      * `“dotted”`

### Supported Time Fields
Users need the following fields to be supported
  * Year
  * Month
  * Week
  * Day
  * Hour
  * Minute
  * Second
  * Millisecond
  * Microsecond
  * Nano Second

#### Design
We are going to support the following fields as a beginning
  * Week
  * Day
  * Hour
  * Minute
  * Second
  * Millisecond
  * Microsecond
  * Nano Second

### Determining Time Fields
Users need to determine the time fields,
 * For example: 6059 seconds
   * Hours, Minutes, Seconds
     * 1 hours, 40 minutes and  59 seconds
   * Minutes and Seconds
     * 100 minutes and 59 seconds.
 
#### Design

>    * Users can determine the time fields using `LargestField` and `SmallestField` parameters.

Reviewer 2: The first letter should be lowercase.  `largestField` and `smallestField`

Author: Good catch !

Done!

   * Example: 
   ```javascript
    {

>       LargestField: "hour",

Reviewer 2: Cite the naming of these fields as an open question: https://github.com/tc39/proposal-temporal/issues/533

Author: I prefer to add the cite to the presentation, what do you think ?

      SmallestField: "second"
    }
  ```
   * Default values
    ```javascript
         LargestField: "Day",
         SmallestField: "second"
    ```
### Hide zero-value fields
- Users needs to hide zero-value fields depends in some criteria
  * Hide all zero-value fields
    * Hours, Minutes, Seconds (without the `hide` option)
      * 1 hours, 0 minutes and 59 seconds.
    * Hours, Minutes, Seconds (with the `hide` option)
      * 1 hours and 59 seconds.
  * Hide-upper-limit with zero-valued.
    * Hours, Minutes, Seconds, MilliSeconds (without the `hide-upper-limit` option)
      * 0 hr, 20 min, 0 sec, 100 msec .
    * Hours, Minutes, Seconds MilliSeconds(with the `hide-upper-limit` option)
      * 20 min, 0 sec, 100 msec.
  * Hide-lower-limit with zero-valued.

#### Design

> - Users can set the `HidingZeroValuedOption` parameter  by one of the following values:

Reviewer 2: If we have this option, I suggest naming it perhaps `hideZeroValues`.  Either way, it should have a lowercase first letter.

Author: Done !

    * `"ALL"` // Hide all the fields that have a zero-valued.

>     * `“LEADING_AND_TRAILING”` // Hide all the zero fields in the leading or the training

Reviewer 2: Make these strings camel case.

`"all"`
`"leadingAndTrailing"`
`"leadingOnly"`
`"trailingOnly"`
`"none"`

See the style guide: https://github.com/tc39/ecma402/blob/master/docs/style-guide.md

Author: Done !

    * `“LEADING_ONLY”`
    * `“TRAILING_ONLY”` 
    * `“NONE”` // Do not hide any zero-valued fields.

- Default Value: `“NONE”`

### Round 
  - Users wants to decide if they are going to round the smallest field or not.
  - for example: 
    * Without rounding option
      * 1 hour and 30.6 minutes.
    * With rounding option

>       * 1 hour and 31 minutes.

Reviewer: Rounding is also something that's being discussed in `Temporal`. My personal opinion is to leave this out from the current draft atleast, but anyway this should be completely dependent on the decision on balancing, since I consider that the same rules would apply here.

Author: I agree with you, but I would like to discuss this next week when we are going to decide where the arithmetic should happened
